modules/navier_stokes/src/materials/NSMushyPorousFrictionMaterial.C

```diff
@@ -0,0 +1,65 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NSMushyPorousFrictionMaterial.h"
+#include "MooseMesh.h"
+#include "NS.h"
+
+registerMooseObject("NavierStokesApp", NSMushyPorousFrictionMaterial);
+
+InputParameters
+NSMushyPorousFrictionMaterial::validParams()
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params.addClassDescription(
+      "Computes the mushy zone porous resistance for solidification/melting problems.");
+  params.addRequiredParam<MooseFunctorName>("liquid_fraction", "Liquid Fraction Functor.");
+  params.addRequiredParam<MooseFunctorName>(NS::mu, "The liquid dynamic viscosity.");
+  params.addRequiredParam<MooseFunctorName>("rho_l", "The liquid density (not the mixture one).");
+  params.addParam<MooseFunctorName>(
+      "dendrite_spacing_scaling", "1e-4", "The dendrite spacing scaling.");
+  return params;
+}
+
+NSMushyPorousFrictionMaterial::NSMushyPorousFrictionMaterial(const InputParameters & parameters)
+  : FunctorMaterial(parameters),
+    _fl(getFunctor<ADReal>("liquid_fraction")),
+    _mu(getFunctor<ADReal>(NS::mu)),
+    _rho_l(getFunctor<ADReal>("rho_l")),
+    _dendrite_spacing_scaling(getFunctor<ADReal>("dendrite_spacing_scaling"))
+{
+
+  addFunctorProperty<ADReal>(
+      "Darcy_coefficient",
+      [this](const auto & r, const auto & t) -> ADReal
+      {
+        constexpr Real epsilon = 1e-15; // prevents explosion of sqrt(x) derivative to infinity
+
+        const auto fs = 1.0 - _fl(r, t);
+        const auto cs = _c / Utility::pow<2>(_dendrite_spacing_scaling(r, t));
+        const auto Fk = 0.5 + std::atan(_s * (fs - _fs_crit)) / pi;
+        const auto K =
+            Utility::pow<3>(_fl(r, t)) / ((Utility::pow<2>(fs) + epsilon) * Fk * cs) + epsilon;
+        return _mu(r, t) / K;
+      });
+
+  addFunctorProperty<ADReal>(
+      "Forchheimer_coefficient",
+      [this](const auto & r, const auto & t) -> ADReal
+      {
+        constexpr Real epsilon = 1e-15; // prevents explosion of sqrt(x) derivative to infinity
+
+        const auto fs = 1.0 - _fl(r, t);
+        const auto cs = _c / Utility::pow<2>(_dendrite_spacing_scaling(r, t));
+        const auto Fk = 0.5 + std::atan(_s * (fs - _fs_crit)) / pi;
+        const auto K =
+            Utility::pow<3>(_fl(r, t)) / ((Utility::pow<2>(fs) + epsilon) * Fk * cs) + epsilon;
+        return _forchheimer_coef * _rho_l(r, t) / std::sqrt(K);
```

Comment:
forchheimer in MOOSE is supposed to contain velocity. Let s double check you have this right when you have time

Reply:
Check `INSFVMomentumFriction.C` the the quadratic coefficient (Forcheimer) already gets multiplied by `std::abs(_u_functor(elem_arg))`

Reply:
so that gives linear with the multiplication there.

do you have u in this coefficient?

Reply:
Sigh. This is why we wanted to delete `INSFVMomentumFriction` in https://github.com/idaholab/moose/pull/20942. It does it differently than `PINSFVMomentumFriction`. BTW if I look at say `FunctorErgunDragCoefficients` in Pronghorn it appears to me that there is no multiplication by velocity in the cQ calculation

Reply:
well to be fair PINSFVmomentumFriction is the one that is wrong. cQ should be just a coefficient, not a function of v

Reply:
I wrote `FunctorErgunDragCoefficients` and its base classes but I just copied the numerics from the non-functor materials

Reply:
Anyway it appears to me that currently the Pronghorn materials are only compatible with the `INSFVMomentumFriction` Forchheimer implementation and are not compatible with the `PINSFVMomentumFriction` implementation

Reply:
So @GiudGiud you're saying you'd like to have a Forchheimer material property that is a product of `cQ` and the velocity?

Reply:
that seems very sensical

Reply:
I think it s the other way, they are only compatible with PINSFV.

No I would like forchhimer to just be cQ.

Reply:
It's so obnoxious that github doesn't auto-update PR conversations. Yea you are right ... the Forchheimer prefactor in the Pronghorn materials includes the speed so they are only compatible with PINSFVMomentumFriction.

So @GiudGiud you want us to change both the material classes and PINSFVMomentumFriction

Reply:
yes. But there's a lot to update. We ll need to fix the manual too

Reply:
How about I add a note to the documentation that this is supposed to be used with `INSFVMomentumFricition` and the person who unifies the whole system can remove it if necessary.

Reply:
This was my task at some point.. I can take it up again

Reply:
> How about I add a note to the documentation that this is supposed to be used with INSFVMomentumFricition and the person who unifies the whole system can remove it if necessary.

This seems reasonable to me. But in that vein should we rename the material (yet again) to `INSFVMushyPorousFrictionMaterial`? We can rename it (with the appropriate registration macro) to `NSFV` after the unification? Changing to INSFV for now may help in preventing users from doing an improper combination

Reply:
I am already an expert at renaming objects, will rename this too.

Reply:
Sounds good for the renaming. Although I must say that incompressible + mushy zone sounds a bit weird lol :)

Reply:
I agree. But in the same vein I found it surprising that the example only used INSFV objects 😉